### PR TITLE
Records grouping

### PR DESCRIPTION
**Pull request by mathemancer**
Related to #3690

Adds a number of SQL functions that produce groupings of records when required. The grouping object submitted should have the keys `columns` and (optionally) `preproc`.

The resulting `grouping` object will have the keys `columns`, `preproc`, and `groups`.

The values associated with each of these keys are the same as in the REST API.

**Technical details**

The new functionality is documented, and if the documentation is insufficient to understand how this works, or the needed request format, please request an improvement.

## Checklist
 
- [X] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [X] My pull request targets the `develop` branch of the repository
- [X] My commit messages follow [best practices][best_practices].
- [X] My code follows the established code style of the repository.
- [X] I added tests for the changes I made (if applicable).
- [X] I added or updated documentation (if applicable).
- [X] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by mathemancer** [comment]
@Anish9901 This can be reviewed, I've just marked it as a draft since it depends on #3708, and should not be merged until after that one.

**Review by Anish9901 — APPROVED** [review]
This looks great @mathemancer, I've requested some minor changes after which this can be merged.

**Comment by Anish9901** [comment]
@mathemancer we should probabaly also metion somewhere that ordering is nessasary for grouping to work as intended.

**Review by mathemancer — COMMENTED** [review]
Review comment so my pending comments will show :hankey: :hankey:

**Comment by mathemancer** [comment]
> @mathemancer we should probabaly also metion somewhere that ordering is nessasary for grouping to work as intended.

I'm not sure. The REST API works the same way, and (in my mind) these operations should be thought of separately. There are other possible UIs that would (for example) just annotate each row with its group ID and/or count.

File: db/sql/test_00_msar.sql [hunk 1]
```diff
@@ -2997,6 +3038,148 @@ END;
...
+BEGIN
+  PERFORM __setup_customers_table();
+  rel_id := '"Customers"'::regclass::oid;
+  RETURN NEXT is(
+    msar.list_records_from_table(
+      rel_id,
+      10, null,
+      '[{"attnum": 3, "direction": "asc"}, {"attnum": 2, "direction": "asc"}]',
+      null,
+      '{"columns": [3, 2]}'
```
[the remark this answers is not in the document]
Comment (mathemancer) [reply db/sql/test_00_msar.sql]: Okay, done.

**Pull request merged**